### ISSUE TITLE
Create application-level remote attestation service

### ISSUE THREAD

**Issue by conradgrobler** [project-oak/oak#1374]
Remote attestation can be done at a protocol level (e.g. EKEP where the attestation is done as part of the connection handshake), but not all clients will be able to support this. For these clients we should provide a gRPC service hosted by the runtime that can be used for the remote attestation as the first call after a connection has been established, but before any data is sent to the server.

The process should be roughly:

- Client establishes TLS connection with the server
- Client generates a random nonce and sends it to the server as the input to the gRPC service call
- The server adds the nonce as additional data, generates a signed attestation report with it and returns it to the client
- The client validates the attestation response before authenticating or having any meaningful interaction with the server

**Comment by benlaurie** [comment]
I think you need to also add some kind of key exchange to prevent MitM...

**Comment by conradgrobler** [comment]
Yes, good point. Does that mean the remote attestation must always be done as part of the connection setup (so always at the protocol level)? It seems that a MitM can always just proxy the key exchange requests as well if they are based on gRPC service calls. Or is there a application-level key exchange approach that could avoid that?

**Comment by conradgrobler** [comment]
Thinking about it some more, it seems that to avoid MitM in this scenario we would have to encrypt all future service calls with the exchanged keys (basically creating a new encrypted channel inside the existing TLS channel), which means it would not be compatible with standard gRPC clients anyway. So this approach does not seem to be a viable option to support a wider range of clients.

**Comment by benlaurie** [comment]
In theory at least, you can change keys on a running TLS connection...

**Comment by tiziano88** [comment]
Maybe a diffie-hellman style exchange (at application layer, on top of TLS) to establish a new shared secret, and then use that as challenge to provide mutual signed attestations? I am not sure that solves the MitM issue though.

cc @daviddrysdale

**Comment by conradgrobler** [comment]
> In theory at least, you can change keys on a running TLS connection...

I was mainly thinking about Andoid clients when creating this issue. I had a quick look, and it does not look like any of the Android TLS security providers compatible with okHttp or cronet exposes any API to allow this. If we have to build a new TLS provider to support this, it it does not seem to provide many benefits over building an EKEP client for Android.

>Maybe a diffie-hellman style exchange (at application layer, on top of TLS) to establish a new shared secret, and then use that as challenge to provide mutual signed attestations? I am not sure that solves the MitM issue though.

I don't think this solves the MitM issue. A MitM can just proxy the key exchange and attestation requests. It won't know the new secret key, but can still facilitate the exchange. If subsequent communications are not encrypted with the new key that was exchanged there is nothing stopping the MitM from evesdropping on or modifying the communications.

**Comment by conradgrobler** [comment]
We should probably reconsider EKEP. I don't think any of the Android TLS providers expose all the API calls we would need to perform and validate a remote attestation over TLS. The same applies to rustls on the gRPC server in the runtime. If we would need to write a new TLS provider for Android and fork rustls to support this, it probably makes sense to rather port EKEP to java and rust.

**Comment by tiziano88** [comment]
I guess this would require at the very least adding EKEP client support for Java (for Android) and Rust (for our own examples).

Also, I wonder whether it will still be possible to do something that works from Chrome, for Oak web applications, using EKEP.
